typing/typetexp.ml

```diff
@@ -234,15 +235,22 @@ and transl_type_aux env policy mode styp =
           let ret_cty = loop acc_mode rest in
           let arg_ty = arg_cty.ctyp_type in
           let arg_ty =
-            if Btype.is_optional l
-            then newty (Tconstr(Predef.path_option,[arg_ty], ref Mnil))
-            else arg_ty
+            if Btype.is_Tpoly (repr arg_ty) then arg_ty else newmono arg_ty
```

Comment:
This code looks suspicious to me.

There are two possibilities: either `repr` can take a non-`Tpoly` type to a `Tpoly` type, or it can't.

* If it can: `Btype.is_Tpoly (repr arg_ty)` might be true, but `arg_ty` might still not be a `Tpoly`, in which case the `Btype.is_mono arg_ty` a few lines down will fail.
* If it can't: Then there's no need to call `repr` here.

Reply:
It was in fact a third possibility `is_mono` calls `repr` itself, whilst `is_Tpoly` doesn't. I called `repr` in `is_mono` because it's generally better to call `repr`. However, I wanted `is_Tpoly` to match the other `is_Tfoo` functions, and they don't call `repr`. Conveniently, upstream has now changed the `is_Tfoo` functions to all call `repr`, so I just backported that change.